hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -81,49 +79,385 @@ class Example(object):
+        if self.index == 0:
+            return None
+        return self.owner.parentage[self.index]
 
-    # Index of example in parent's children, or None if this is the root.
-    child_index = attr.ib(default=None)
+    @property
+    def start(self):
+        """The position of the start of this example in the byte stream."""
+        return self.owner.starts[self.index]
+
+    @property
+    def end(self):
+        """The position directly after the last byte in this byte stream.
+        i.e. the example corresponds to the half open region [start, end).
+        """
+        return self.owner.ends[self.index]
 
-    # An example is "trivial" if it only contains forced bytes and zero bytes.
-    # All examples start out as trivial, and then get marked non-trivial when
-    # we see a byte that is neither forced nor zero.
-    trivial = attr.ib(default=True, repr=False)
+    @property
+    def depth(self):
+        """Depth of this example in the example tree. The top-level example has a
+        depth of 0."""
+        return self.owner.depths[self.index]
 
-    # True if we believe that the shrinker should be able to delete this
-    # example completely, without affecting the value produced by its enclosing
-    # strategy. Typically set when a rejection sampler decides to reject a
-    # generated value and try again.
-    discarded = attr.ib(default=None, repr=False)
+    @property
+    def trivial(self):
+        """An example is "trivial" if it only contains forced bytes and zero bytes.
+        All examples start out as trivial, and then get marked non-trivial when
+        we see a byte that is neither forced nor zero."""
+        return self.index in self.owner.trivial
 
-    # List of child examples, represented as indices into the example list.
-    children = attr.ib(default=attr.Factory(list), repr=False)
+    @property
+    def discarded(self):
+        """True if this is example's ``stop_example`` call had ``discard`` set to
+        ``True``. This means we believe that the shrinker should be able to delete
+        this example completely, without affecting the value produced by its enclosing
+        strategy. Typically set when a rejection sampler decides to reject a
+        generated value and try again."""
+        return self.index in self.owner.discarded
 
-    # We access length a lot, and Python is annoyingly bad at basic integer
-    # arithmetic, so it makes sense to cache this on a field for speed
-    # reasons. It also reduces allocation, though most of the integers
-    # allocated from this should be easily collected garbage and/or
-    # small enough to be interned.
-    length = attr.ib(init=False, repr=False)
+    @property
+    def length(self):
+        """The number of bytes in this example."""
+        return self.end - self.start
+
+    @property
+    def child_index(self):
+        """The index of the current example in its parents children list."""
```

Comment:
`parents` → `parent's`

I also think “parent's list of children” would be a little clearer here.

(And perhaps `index_in_parent` is a clearer name for the property itself, since `child_index` sounds related to the *current* node's children.)

Reply:
Actually, lets review/merge #1870 first and just delete the `child_index` code as no longer needed.